### ISSUE TITLE
FH Challenges (Spoiler!)

### ISSUE THREAD

**Issue by Lurkars** [Lurkars/gloomhavensecretariat#96]
Implement FH challenges feature, at least what's possible. Should be easily adopted with the scenario rules. Addition may need to change monster stats (abilities). Data available at https://github.com/any2cards/worldhaven/tree/master/images/challenge-cards/frosthaven

**Comment by z-m-k** [comment]
Even if not fully implemented, can you add a toggle to reflect that a group has X active challenges and award 2X experience to each player? I can see this done the following way: add a toggle that challenges are unlocked and then on the 'success' screen reuse the 'goal' tick boxes to add a 'challenge' line.

**Comment by Lurkars** [comment]
I don't have access to the rule book stickers yet, so don't have the challenge rules available. Can you provide me those or give more details?

**Comment by Lurkars** [comment]
Ah okay, I found some info, so important is only the rule "each character gains 2XP for every challenge played on success". Any good idea how to add this feature spoiler free? I mean I personally do not care, but unsure if I just add a challenge field to the success summary, if this is spoiler for some folks.

**Comment by TheSquinter** [comment]
Could be just a tick box in a corner of the summary screen or below battle goals called 'Challenge' or 'Challenge Compete'? Ticking it adds +2 bonus XP to all characters not absent. Most people would know of the challenge deck in general due to the setting up, but not know the details.

Or can you make it hidden behind spoiler text like "Building #" that you need to click on to view the words "Challenge Complete"

**Comment by Lurkars** [comment]
Damn in the end would be best if I had #163 finished and make it depended from that. But still no good idea where and how to add buildings support. The general data is already there.

**Comment by Lurkars** [comment]
So now with tracking of buildings in #163, I hope I can quickly add this in the basic function to the scenario summary.

**Comment by Lurkars** [comment]
So in [0.62.4](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.62.4) I add challenges if Town Hall is build. On Scenario summary there is a checkbox to mark if the challenge was successful, in building level 3 there are two checkboxes of course. I hope this is enough for the start.

**Comment by TheSquinter** [comment]
Did a quick test, seems to work but there might be a bug with forced linked scenarios:
- Manually added Building 90 and built it
- Started scenario 2 and clicked Success
- Challenges tickbox appears on Scenario summary and ticking it shows the +2 Challenge XP
- Clicking "Continue with #4B" only added the 6 bonus XP, but not the +2 Challenge XP

Same with non-forced linked scenarios where you can either finish or continue with link

![image](https://user-images.githubusercontent.com/124784924/235017564-3b7e1141-3779-4ea3-8a5d-1399e7cd0ec9.png)

**Comment by Lurkars** [comment]
Thanks, still check.

**Comment by Lurkars** [comment]
Should be fixed in  [v0.62.5](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.62.5)

**Comment by Lurkars** [comment]
Town Guard Deck is also added a few version ago (since [v0.65.4](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.65.4)). It's available on the campaign sheet as slider:


![Screenshot from 2023-05-23 21-33-05](https://github.com/Lurkars/gloomhavensecretariat/assets/2760570/3eb00d52-04dc-4190-91ee-5d92c2374e18)
![Screenshot from 2023-05-23 21-33-19](https://github.com/Lurkars/gloomhavensecretariat/assets/2760570/20546838-cf2f-44a4-b7e9-81acb075deea)

Perks are already supported. What's missing by now is as I see that some section/rewards or so also add cards to this, this is currently missing. Also keep in mind, that drawing is only possible when not playing a scenario. And maybe it need to be shuffled manually, haven't re-read the rules yet, how it completely works.

**Comment by StaffanBaloo** [comment]
Regarding challenges, I think some of them make global adjustments to monster stats. For example, I saw someone on Reddit mention a challenge that gave all monsters +2 health.
I figure that implementing the challenge deck itself is probably a lot of work, but I'm guessing there is already support for scenario-based stat adjustments (since there are some scenarios that do things like "all monsters add push 2 to their attacks"). Perhaps it would be possible to allow manual access to these modifiers and handle it that way? Or are the modified monsters implemented as their own set of monsters (e.g. there's both a "Snow Imp" and a "Snow Imp (scenario X)" in the data files?

**Comment by Lurkars** [comment]
For now, special scenario rules are technically complete new monsters. So this is not an easy change. Anyways, since there are even more use cases for this, I am already thinking about a better format to "change" existing monsters dynamically and use that also for the scenario rules. But still on a long backlog list, but of course, this ticket is open to add support for this sometimes and that will of course include the stuff you mentioned. But as already mentioned, by now the monsters have different files (simple example [City Guard from Scenario 1](https://github.com/Lurkars/gloomhavensecretariat/blob/main/data/fh/monster/city-guard-scenario-1.json))

**Comment by Lurkars** [comment]
first draft in [v0.99.13](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.99.13):
- [x] create shuffled deck when Town Hall Build
- [x] draw cards before starting first round (depending on TH level)
- [x] choose cards to keep active for scenario (depending on TH level)

Whats currently missing:
- [ ] automatic apply some card rules

**Comment by sacrelicious2** [comment]
How do you enable the challenge cards?  I don't see any such options under Frosthaven Rules

**Comment by Lurkars** [comment]
So finally this can be moved to a testing state in [v0.102.2](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.102.2). I tried to implement all automatons possible.

Happy if some can also help testing here, because I am still not in a FH Campaign with our group and even if we start this year, will take some time until we unlock challenges.

**Comment by tttopuz** [comment]
What's needed to test here? Maybe I can assist.

**Comment by Lurkars** [comment]
> What's needed to test here? Maybe I can assist.

Thanks for your support! Just general using the challenge deck if it is working as expected. I just check functionality, but not the flow in actual gameplay. Sometimes this shows up some quirks etc.

**Comment by alibaba56734** [comment]
You're right. I was testing the challenges before draw of the first turn. After first drawing, it is not possible to draw cards any more. Maybe a reminder would be nice to set challenges if no challenges are active before drawing first turn.
